### PR TITLE
Add precondition ifGenerationMatch=0 for create directory calls

### PR DESCRIPTION
**Pull request by mprashanthsagar**

### CODE REVIEW

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java [hunk 1]
```diff
@@ -1965,10 +1965,11 @@ private StorageObject getObject(StorageResourceId resourceId)
   private boolean canIgnoreExceptionForEmptyObject(
       IOException exceptionOnCreate, StorageResourceId resourceId, CreateObjectOptions options)
       throws IOException {
-    // TODO(user): Maybe also add 409 and even 412 errors if they pop up in this use case.
+    // TODO(user): Maybe also add 409 errors if they pop up in this use case.
     // 500 ISE and 503 Service Unavailable tend to be raised when spamming GCS with create requests:
     if (errorExtractor.rateLimited(exceptionOnCreate)
-        || errorExtractor.internalServerError(exceptionOnCreate)) {
+        || errorExtractor.internalServerError(exceptionOnCreate)
+        || errorExtractor.preconditionNotMet(exceptionOnCreate)) {
```
Comment (medb) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java:1972]: Probably we need ignore this only for directories?

File: gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemNewIntegrationTest.java [hunk 2]
```diff
@@ -101,11 +120,62 @@ public void mkdirs_shouldCreateNewDirectory() throws Exception {
     String dirObject = getTestResource();
     URI dirObjectUri = new URI("gs://" + bucketName).resolve(dirObject);
 
+    gcsfsIHelper.mkdir(
+        bucketName, dirObject); // create directory before hand without tracking requests
```
Comment (medb) [inline gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemNewIntegrationTest.java:124]: Move comment on before the `mkdir` call and re-format code.

File: gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemNewIntegrationTest.java [hunk 3]
```diff
@@ -101,11 +120,62 @@ public void mkdirs_shouldCreateNewDirectory() throws Exception {
...
+    gcsfsIHelper.mkdirs(dirObjectUri); // create directory before hand without tracking requests
+    gcsFs.mkdirs(dirObjectUri);
+
+    assertThat(gcsRequestsTracker.getAllRequestStrings())
+        .containsExactly(
+            batchRequestString(),
+            getRequestString(bucketName, dirObject),
+            getRequestString(bucketName, dirObject + "/d1"),
+            uploadRequestString(bucketName, dirObject + "/d1/", /* generationId= */ 1),
+            getRequestString(bucketName, dirObject + "/d1/")); // verifies directory exists
```
Comment (medb) [inline gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemNewIntegrationTest.java:178]: Move comment on new line before command that it refers.

File: gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemNewIntegrationTest.java [hunk 4]
```diff
@@ -101,11 +120,62 @@ public void mkdirs_shouldCreateNewDirectory() throws Exception {
...
     URI dirObjectUri = new URI("gs://" + bucketName).resolve(dirObject);
 
+    gcsfsIHelper.mkdir(
+        bucketName, dirObject); // create directory before hand without tracking requests
     gcsFs.mkdir(dirObjectUri);
 
     assertThat(gcsRequestsTracker.getAllRequestStrings())
         .containsExactly(
-            uploadRequestString(bucketName, dirObject + "/", /* generationId= */ null));
+            uploadRequestString(bucketName, dirObject + "/", /* generationId= */ 1),
```
Comment (medb) [inline gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemNewIntegrationTest.java:129]: Probably we would want to check that generation ID was `0` in this case. Maybe by adding `getAllRawRequestStrings` and adding overloaded `uploadRequestString` method with `replaceGenerationId` parameter, like here: https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/7891c3d898944c7f47b70fe4e7d276bc4da5c3c1/gcsio/src/test/java/com/google/cloud/hadoop/gcsio/TrackingHttpRequestInitializer.java#L259

**Review by medb — CHANGES_REQUESTED** [review]

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java (hunk 1 above)
Comment (mprashanthsagar) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java:1972]: Done

File: gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemNewIntegrationTest.java (hunk 2 above)
Comment (mprashanthsagar) [reply gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemNewIntegrationTest.java:124]: Done

File: gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemNewIntegrationTest.java (hunk 4 above)
Comment (mprashanthsagar) [reply gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemNewIntegrationTest.java:129]: Done

File: gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemNewIntegrationTest.java [hunk 5]
```diff
@@ -103,9 +103,84 @@ public void mkdirs_shouldCreateNewDirectory() throws Exception {
 
     gcsFs.mkdir(dirObjectUri);
 
-    assertThat(gcsRequestsTracker.getAllRequestStrings())
+    assertThat(gcsRequestsTracker.getAllRawRequestStrings())
+        .containsExactly(
+            uploadRequestString(bucketName, dirObject + "/", /* generationId= */ 0, false));
```
Comment (medb) [inline gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemNewIntegrationTest.java:108]: Add a comment for `false` argument, here and below:

```java
/* replaceGenerationId= */ false,
```

**Review by medb — APPROVED** [review]

**Pull request merged**